src/dns.rs

```diff
@@ -0,0 +1,464 @@
+use crate::topology::config::GlobalOptions;
+use futures::{Async, Future, Poll};
+use hyper::client::connect::dns::{Name, Resolve};
+use std::io;
+use std::net::ToSocketAddrs;
+use std::net::*;
+use std::vec::IntoIter;
+use trust_dns_resolver::config::*;
+use trust_dns_resolver::error::ResolveErrorKind;
+use trust_dns_resolver::AsyncResolver;
+use trust_dns_resolver::BackgroundLookupIp;
+
+/// Default port for DNS service.
+const DNS_PORT: u16 = 53;
+
+/// Default version will behave identically to std::net::ToSocketAddrs implementations.
+#[derive(Clone, Default)]
+pub struct DnsResolver(Option<AsyncResolver>);
+
+impl DnsResolver {
+    /// (Self,errors)
+    pub fn new(config: &GlobalOptions) -> Result<Self, String> {
```

Comment:
We should probably create an actual error type like we do with our sinks sources etc

Reply:
I think it should be like this, the reason being that the main call site in builder::build_pieces expects errors in String form.

Reply:
Yeah I assume that will change in the future, and we should keep the string conversion at the topology config level instead of deep within a util piece like this. Because you can always convert any `std::error::Error` into a string.

Reply:
Good point